openff/system/components/potentials.py

```diff
@@ -31,20 +30,29 @@ def validate_parameters(cls, v):
 class PotentialHandler(DefaultModel):
     """Base class for storing parametrized force field data"""
 
-    name: str
-    expression: str
-    independent_variables: Union[str, Set[str]]
-    slot_map: Dict[TopologyKey, PotentialKey] = dict()
-    potentials: Dict[PotentialKey, Potential] = dict()
-
-    # Pydantic silently casts some types (int, float, Decimal) to str
-    # in models that expect str; this may be updates, see #1098
-    @validator("expression", pre=True)
-    def is_valid_expr(cls, val):
-        if isinstance(val, str):
-            return val
-        else:
-            raise InvalidExpressionError
+    type: str = Field(..., description="The type of potentials this handler stores.")
+    expression: str = Field(
+        ...,
+        description="The analytical expression governing the potentials in this handler.",
+    )
+    slot_map: Dict[TopologyKey, PotentialKey] = Field(
+        dict(),
+        description="A mapping between TopologyKey objects and PotentialKey objects.",
+    )
+    potentials: Dict[PotentialKey, Potential] = Field(
+        dict(),
+        description="A mapping between PotentialKey objects and Potential objects.",
+    )
+
+    @property
+    def independent_variables(self) -> Set[str]:
+        vars_in_potentials = set([*self.potentials.values()][0].parameters.keys())
+        vars_in_expression = {
+            node.id
+            for node in ast.walk(ast.parse(self.expression))
+            if isinstance(node, ast.Name)
+        }
+        return vars_in_expression - vars_in_potentials
```

Comment:
This block of code (self-contained) might be of interest to GMSO developers (@justinGilmer @umesh-timalsina @daico007)

```
In [24]: import ast
    ...:
    ...:
    ...: def get_vars_in_expr(expr):
    ...:     names = {
    ...:         node.id for node in ast.walk(ast.parse(expr))
    ...:         if isinstance(node, ast.Name)
    ...:     }
    ...:     return names
    ...:

In [25]: get_vars_in_expr('m*x+b')
Out[25]: {'b', 'm', 'x'}

In [26]: get_vars_in_expr("4*epsilon*((sigma/r)**12-(sigma/r)**6)")
Out[26]: {'epsilon', 'r', 'sigma'}
```

Taken from https://stackoverflow.com/a/41038615/4248961

Reply:
@mattwthompson Thank You for pointing this out. We currently have a problem with loading large [forcefield](https://github.com/mosdef-hub/gmso/issues/521) files. Validating potential expressions takes the longest amount of time. We have discussed some memoizing ideas. Together with the approach mentioned here and memoization, it looks like we can get performance benefits.